### PR TITLE
Virtual tickets for Fall Carnival 2023

### LINKED ISSUE #430: Fall carnival: Virtual tickets
Implement #409 for the Fall Carnival 2023

# Views/routes

- `/events/fall-carnival-2023` Event page with editor (editor requires log in)
- `/events/fall-carnival-2023/tickets/hellsatanx` Customized ticket url to share

# Questions

- Do we need to check if users are banned and not allow banned users to create tickets?

# MVP

- Assets:
    - Ticket
       - [x] Base layer
       - [x] Mask
       - [x] Overlay (Text "Alveus Fall Carnival", "Nov. 4th 2023", "twitch.tv/maya")
       - [ ] attribute sticker artists on the ticket
	- [ ] Sticker set "Fall carvinal"
		- [x] pumkin(s)
		- [x] emojiconmotes (ambassadors)?
		- [x] more ideas are welcome Smile
- Ticket editor
   - [x] Place stickers (keep in bounds)
   - [x] Move stickers
   - [x] Remove stickers
   - [x] Support mobile and desktop
   - [x] reset button
   - [x] group stickers (by type/art style)
   - [x] attribute artists for stickers
- [x] DB table (`user_id`, `event`, `signup_at`, `customized_ticket_data`, ...)
- [x] Serialize custom ticket (-> save to db)
- [x] Rasterize custom ticket (-> save jpeg/png on cdn)
- [ ] QR-Code / Overlay for stream "Claim your custom Alveus Fall Carnival Ticket"
- [ ] Implement event page
    - [x] "Alveus Fall Carvinal, November 4th 2023, on twitch.tv/maya"
    - [x] More info about the event? Link to other events?
    - [x] not logged in -> "Log in with Twitch to claim a ticket"
    - [x] logged in -> "Customize a ticket"
- [ ] Implement ticket customization page
    - [x] Embed editor (see above)
    - [x] Share Button (Link-URL, Socials, E-Mail?)
    - [x] Download (-> jpeg/png)
- [ ] Implement ticket view
    - [x] Either load bitmap from cdn or render ticket from db
    - [x] Support og to share on socials. Make sure the image loads fast!
    - [x] if not customized -> "This person has not claimed their ticket yet" or "404 not found"
    - [x] if customized -> show readonly ticket of that person
    - [x] Repeat the buttons from the event page (Log in & Claim your own ticket / Customize your own ticket)
    - [x] On mobile: either allow to zoom the ticket on mobile or make sure it does at least work okay on vertical smartphone screens?!

# Ideas

- Event page
    - [ ] Add to calendar button?
    - [ ] Remind me -> Notification Bell Button?
- [ ] Generate QR-Code for customized ticket (`https://alveus.gg/events/fall-carnival-2023/tickets/{username}`)
- Ticket customization page
    - [ ] Maybe a delete button?
- More pages
    - [ ] my tickets (see #42 )
    - [ ] gallery of the tickets generated for one event
- Editor features:
    - [ ] scale ticket on mobile (pinch to zoom) to make placing stickers easier
    - [ ] resize stickers
    - [ ] rotate stickers
    - [ ] use stickers multiple times
    - [ ] move stickers in z-index
    - [ ] drag stickers from the pack to the ticket

### PR DESCRIPTION
**Pull request by pjeweb**
## Describe your changes

Resolves #430 

### Implementation notes

- Using framer motion and the DOM to edit the stickers
- Using satori in app router edge route handlers to render the tickets on the server
   - this means we cant use all of tailwind and to make it easier i am just using inline styles for the few components that get rendered in the tickets
   - had to patch satori to support mask image on elements with positions other than `0,0` => https://github.com/vercel/satori/issues/565
   - we have to serve the asset (background, foreground and stickers from directly and load via http - no imports), and since we share the code with the client we simple ``s there too
- Using planetscale's edge http connection to access the ticket data on the edge function -> not compatible with other mysql hosts
- Using blocking SSR for the sharable URL each ticket gets so the `og:image` can get read by the social networks

### Bugs

- [x] finished ticket disappears on iOS
- [x] purge ticket cache on changes
- [x] fix og:image meta data

## Notes for testing your change

- test embed on social sharing (does not work on preview.alveus.gg)

### CODE REVIEW

File: apps/website/package.json [hunk 1]
```diff
@@ -35,6 +35,7 @@
     "@next-auth/prisma-adapter": "^1.0.7",
     "@next/env": "^13.5.1",
     "@paralleldrive/cuid2": "^2.2.2",
+    "@planetscale/database": "^1.11.0",
```
Comment (pjeweb) [inline apps/website/package.json:38]: Note: Using this database access through http on edge functions will only work with planetscale and not with other mysql hosters. So we probably should look for another solution instead. Just used it 'cause its easy and has very low runtime cost.

File: apps/website/src/app/api/virtual-tickets/[event]/[user]/og/route.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import { ImageResponse } from "@vercel/og";
```
Comment (pjeweb) [inline apps/website/src/app/api/virtual-tickets/[event]/[user]/og/route.tsx:1]: Ideally the two handlers should be DRYed, and abstracted to be usable for future events.

File: apps/website/src/components/content/Meta.tsx [hunk 3]
```diff
@@ -21,8 +21,6 @@ const Meta = ({ title, description, image, children }: MetaProps) => {
     "Alveus is a 501(c)(3) nonprofit organization functioning as a wildlife sanctuary & virtual education center following the journeys of our non-releasable ambassadors, aiming to educate and spark an appreciation for them and their wild counterparts.";
   const computedTitle = title ? `${title} | ${defaultTitle}` : defaultTitle;
   const computedDescription = description || defaultDescription;
-
-  // Use the Next.js image loader, and use an absolute URL
```
Comment (pjeweb) [inline apps/website/src/components/content/Meta.tsx:25]: TODO: add the comment back

File: apps/website/src/components/events/fall-carnival-2023/IntroSection.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        <div className="mb-2 inline-block rounded-lg bg-white/90 px-2 py-0.5 text-xs uppercase text-[#4E1362]">
+          Upcoming event
+        </div>
+
+        <Heading
+          style={{
+            // NOTE: react ts typings don't know about text-wrap yet so we have to disable the rule
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
+            textWrap: "balance",
```
Comment (pjeweb) [inline apps/website/src/components/events/fall-carnival-2023/IntroSection.tsx:53]: Do we want to add this as a default to headings or at least configure it to be used in tailwind? Browser support is still pretty sparse, but its such a nice feature https://caniuse.com/css-text-wrap-balance

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+            ...state,
+            dirty: true,
+            stickers: [],
+          };
+        case "update":
+          return { ...state, dirty: true };
+        case "save":
+          return { ...state, dirty: false };
+      }
+    },
```
Comment (pjeweb) [inline apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:99]: TODO: Move reducer outside of the component

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    [height, stickerPack, width],
+  );
+
+  const resetStickers = useCallback(() => {
+    stickerPositions.current = [];
+    dispatch({ type: "reset" });
+  }, []);
+
+  return {
+    isDirty: state.dirty,
```
Comment (pjeweb) [inline apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:184]: memo the return object?

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    stickers: state.stickers,
+    addSticker,
+    removeSticker,
+    updateStickerPosition,
+    stickerPositionsRef: stickerPositions,
+    setStickerData,
+    saveStickers: () => {
+      dispatch({ type: "save" });
+    },
```
Comment (pjeweb) [inline apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:193]: memo the callback?

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+          <IconTrash className="h-6 w-6" />
+          Reset
+        </Button>
+      </div>
+
+      <div className="mx-auto mt-8 flex max-w-3xl flex-row gap-4">
+        {typeSafeObjectKeys(stickerPack.groups).map((groupId) => {
+          const group = stickerPack.groups[groupId]!;
+
+          return (
```
Comment (pjeweb) [inline apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:299]: maybe refactor a bit so its not nested that many levels. maybe extract a component

File: apps/website/src/components/events/virtual-ticket/Wiggle.tsx [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        };
+
+        const distanceToMouse = Math.sqrt(
+          Math.pow(event.clientX - cardCenter.x, 2) +
+            Math.pow(event.clientY - cardCenter.y, 2),
+        );
+        const averageRadius = (cardRect.width + cardRect.height) / 4;
+
+        // Calculate intensity, clamped between 0 and 1
+        const lightIntensity = Math.min(distanceToMouse / averageRadius, 1);
```
Comment (pjeweb) [inline apps/website/src/components/events/virtual-ticket/Wiggle.tsx:72]: comments don't make sense

File: apps/website/src/pages/events/fall-carnival-2023/index.tsx [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    isDirty,
+    stickerPositionsRef,
+    setStickerData,
+    updateStickerPosition,
+    saveStickers,
+  } = stickerData;
+
+  useEffect(() => {
+    if (myTicket.data) setStickerData(myTicket.data.customization.stickers);
+  }, [myTicket.data, setStickerData]);
```
Comment (pjeweb) [inline apps/website/src/pages/events/fall-carnival-2023/index.tsx:64]: using useEffect feels wrong but we somehow need to sync the data from the server=trpc to the live positions updated by framer motion. there probably is a way to directly pass the trpc/react-query data to the sticker components and update the data from there but i could not figure it out in time

File: apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+export const getStaticPaths = (async () => {
+  return {
+    paths: [],
+    fallback: "blocking",
+  };
+}) satisfies GetStaticPaths;
+
+export const getStaticProps = (async () => ({
+  props: {},
+})) satisfies GetStaticProps;
```
Comment (pjeweb) [inline apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx:34]: Add a comment explaining why we need this: We use blocking SSR cause we want to ensure each page gets its own unique metadata (`og:image`) for sharing the links. We don't know the username in advance so it cant be statically generated.

File: apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+  const ticket = trpc.virtualTickets.getTicket.useQuery(
+    { eventId, userName: userName as string },
+    { enabled: !!userName, retry: 2 },
+  );
+
+  const showTicket = userName && ticket.data;
+
+  const ogImage =
+    userName &&
```
Comment (pjeweb) [inline apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx:51]: Instead of repeatedly checking if the username is set, maybe we can use `invariant` to check instead.

File: apps/website/src/server/bingo-cards.ts
Comment (pjeweb) [inline apps/website/src/server/bingo-cards.ts:1]: TODO: this file does not belong in this PR

File: apps/website/src/server/bingo-distribution.ts
Comment (pjeweb) [inline apps/website/src/server/bingo-distribution.ts:1]: TODO: this file does not belong in this PR

File: patches/next@13.5.4.patch
Comment (pjeweb) [inline patches/next@13.5.4.patch:1]: This patch is needed for the masking until https://github.com/vercel/satori/pull/566 gets merged or https://github.com/vercel/satori/issues/565 get fixed.

File: patches/next@13.5.4.patch [hunk 13]
```diff
@@ -0,0 +1,26 @@
+diff --git a/dist/compiled/@vercel/og/index.edge.js b/dist/compiled/@vercel/og/index.edge.js
```
Comment (MattIPv4) [inline patches/next@13.5.4.patch:1]:
```suggestion
# FIXME: https://github.com/vercel/satori/issues/565
diff --git a/dist/compiled/@vercel/og/index.edge.js b/dist/compiled/@vercel/og/index.edge.js
```

File: apps/website/src/components/layout/Layout.tsx [hunk 14]
```diff
@@ -78,6 +85,18 @@ const Layout = ({ children }: LayoutProps) => {
...
+        {showTopHat && (
+          <Link
+            href="/events/fall-carnival-2023"
+            className="relative z-10 block border-b border-b-[#28122f] bg-[#4E1362] px-4 pb-1 pt-1.5 text-sm text-white"
+          >
+            <div className="container mx-auto flex items-center justify-center gap-1">
+              Alveus Fall Carnival &middot; Nov. 4th &middot; Get your ticket
+              <IconArrowRight className="h-4 w-4" />
+            </div>
+          </Link>
```
Comment (MattIPv4) [inline apps/website/src/components/layout/Layout.tsx:97]: Can we add a hover state to this, please?

File: apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+            </Wiggle>
+
+            <p className="text-center">
+              Sticker art by Mik_MWP (Sub emotes) <br />
+              and{" "}
+              <Link external href="https://www.instagram.com/lazygoosepixels/">
+                LazyGoose
+              </Link>{" "}
+              (Pixel Ambassadors)
+            </p>
```
Comment (MattIPv4) [inline apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx:135]: Can this be generated from the sticker pack config, instead of being hard-coded here?

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+      }
+    }
+  }, []);
+
+  return (
+    <ticketEditorContext.Provider value={contextValue}>
+      <div className="relative -mx-4 w-[calc(100%+2rem)] bg-black/10">
+        <div className="pointer-events-none absolute left-0 right-0 top-0 mt-1 flex flex-row items-center justify-center gap-2 p-2 text-black/50 lg:hidden">
+          <IconChevronLeft className="h-4 w-4" />
+          scroll here to move the ticket
```
Comment (MattIPv4) [inline apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:269]:
```suggestion
          swipe to move ticket
```

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx (hunk 8 above)
Comment (MattIPv4) [reply apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:342]: Agree, would probably be nice to have a StickerPack component, could just be in this file

File: apps/website/src/components/events/virtual-ticket/TicketEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+                      {group.attribution}
+                    </Link>
+                  ) : (
+                    group.attribution || <>&nbsp;</>
+                  )}
+                </p>
+              </div>
+
+              <div className="w-full overflow-auto md:overflow-hidden">
+                <ul className="flex flex-row items-center gap-1 py-2 md:flex-wrap md:gap-2 lg:gap-3">
```
Comment (MattIPv4) [inline apps/website/src/components/events/virtual-ticket/TicketEditor.tsx:358]: I think these should probably wrap on mobile too, rather than using scroll -- I'm not sure the fact you can scroll for more is very obvious on mobile

**Review by MattIPv4 — COMMENTED** [review]
Is there a reason the stickers are in public, rather than standard assets that get included with the bundle?

---

It might be good to add subdirectories to where the stickers are, grouping them by the sticker pack? It's a bit confusing having the emotes combined with the pixel ones (albeit they have a prefix), and the random halloween stickers too.

---

Long term I think it'd be really cool to move the sticker sets to the data repo, and have them tied to the ambassadors? So from Georgie's ambassador entry, you could get his emote + pixel stickers?

For example, we could then fetch the pixel sticker for each ambassador and use it as the favicon on each ambassador's page.

**Comment by pjeweb** [comment]
> Is there a reason the stickers are in public, rather than standard assets that get included with the bundle?

As far as i can see satori/next needs it to be static files in the public path for the edge handler to access the images through http. Its what all examples show. At least I could not figure out how to use it with the image loader through imports.

> It might be good to add subdirectories to where the stickers are, grouping them by the sticker pack? It's a bit confusing having the emotes combined with the pixel ones (albeit they have a prefix), and the random halloween stickers too.

agreed and done

> Long term I think it'd be really cool to move the sticker sets to the data repo, and have them tied to the ambassadors? So from Georgie's ambassador entry, you could get his emote + pixel stickers?
>
> For example, we could then fetch the pixel sticker for each ambassador and use it as the favicon on each ambassador's page.

sounds good yeah

File: apps/website/src/assets/fonts/PT_Sans/OFL.txt
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/assets/fonts/PT_Sans/OFL.txt:1]: Same problem as with the stickers and assets for satori in general. We might be able to copy them as a build step from those packages. Not sure if thats worth?

File: apps/website/src/assets/fonts/PT_Serif/OFL.txt
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/assets/fonts/PT_Serif/OFL.txt:1]: dito

File: apps/website/src/assets/fonts/PT_Sans/OFL.txt
[the remark this answers is not in the document]
Comment (MattIPv4) [reply apps/website/src/assets/fonts/PT_Sans/OFL.txt:1]: Hm, but this is in src/assets, not public/assets? I think it'd be better to use fontsource rather than having our own copies, and write a quite copy if needed

[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/assets/fonts/PT_Sans/OFL.txt:1]: Oh yeah I guess i don't understand how fetch/URL works in that context. Next seems to be bundling files fetched like that, even if its from outside of src or public.

File: apps/website/src/assets/fonts/PT_Serif/OFL.txt
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/assets/fonts/PT_Serif/OFL.txt:1]: just removing the dep as we dont use it in the tickets (yet)

File: apps/website/src/pages/events/fall-carnival-2023/index.tsx [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+            <div className="flex-1">
+              <LoginWithTwitchButton />
+            </div>
+          </div>
+        )}
+
+        {session.status === "authenticated" && userName && (
+          <>
+            <div className="relative my-8 flex flex-row items-center gap-4 align-middle">
+              <div className={isDirty ? "pointer-events-none opacity-40" : ""}>
```
Comment (MattIPv4) [inline apps/website/src/pages/events/fall-carnival-2023/index.tsx:122]: classes helper, or:

```suggestion
              <div className={isDirty ? "pointer-events-none opacity-40" : undefined}>
```

File: apps/website/src/pages/events/fall-carnival-2023/index.tsx [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+                <a
+                  className={`flex flex-row items-center justify-center gap-1 rounded-xl bg-gray-700/10 p-1 px-3 text-center hover:bg-gray-700/20`}
```
Comment (MattIPv4) [inline apps/website/src/pages/events/fall-carnival-2023/index.tsx:124]:
```suggestion
                  className="flex flex-row items-center justify-center gap-1 rounded-xl bg-gray-700/10 p-1 px-3 text-center hover:bg-gray-700/20"
```

File: apps/website/src/pages/events/fall-carnival-2023/index.tsx [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+                  download="alveus-fall-carnival-2023-ticket.png"
+                  target="_blank"
+                  rel="download"
+                >
+                  <IconDownload className="mr-1 h-5 w-5" />
+                  Download
+                </a>
+              </div>
+
+              <div className={isDirty ? "pointer-events-none opacity-40" : ""}>
```
Comment (MattIPv4) [inline apps/website/src/pages/events/fall-carnival-2023/index.tsx:135]: Ditto, helper or
```suggestion
              <div className={isDirty ? "pointer-events-none opacity-40" : undefined}>
```

File: apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+      <div className="-mt-40 hidden h-40 bg-carnival bg-gradient-to-t from-carnival to-carnival-700 lg:block" />
+
+      <div className="relative">
+        <Image
+          src={leafLeftImage2}
+          alt=""
+          className="pointer-events-none absolute -bottom-24 left-0 z-10 hidden h-auto w-1/2 max-w-[12rem] select-none lg:block"
+        />
+
+        <IntroSection showLinkToClaimTicket={true} />
```
Comment (MattIPv4) [inline apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx:99]:
```suggestion
        <IntroSection showLinkToClaimTicket />
```

File: apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+            >
+              <div className="absolute inset-0 flex items-center justify-center bg-carnival text-2xl font-bold text-white">
+                {!showTicket && <span>Loading ticket…</span>}
+              </div>
+
+              {/* eslint-disable-next-line @next/next/no-img-element */}
+              <img
+                className={`relative w-full opacity-0 transition-opacity ${
+                  showTicket ? "block opacity-100" : "hidden"
+                }`}
```
Comment (MattIPv4) [inline apps/website/src/pages/events/fall-carnival-2023/tickets/[userName].tsx:126]: classes helper

File: apps/website/src/app/api/virtual-tickets/[event]/[user]/ticket/route.tsx [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+      return new Response("Could not parse customization", { status: 404 });
+    }
+
+    const stickers = customization.data.stickers;
+
+    const fontPtSansBold = await fetch(
+      new URL(
+        "../../../../../../../node_modules/@fontsource/pt-sans/files/pt-sans-latin-700-normal.woff",
+        import.meta.url,
+      ),
```
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/app/api/virtual-tickets/[event]/[user]/ticket/route.tsx:52]: does not look like it. i get this error `{}.resolve is not a function`

File: apps/website/src/components/events/fall-carnival-2023/VirtualTicket.tsx [hunk 24]
```diff
@@ -0,0 +1,77 @@
+import { ticketConfig } from "@/data/events/fall-carnival-2023";
+
+import { getShortBaseUrl } from "@/utils/short-url";
+
+import { QRCode } from "@/components/QrCode";
+
+import { Ticket, type TicketProps } from "../virtual-ticket/Ticket";
```
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/components/events/fall-carnival-2023/VirtualTicket.tsx]: not sure. @ feels slightly better here. maybe only on the same level? i dont really care much

File: apps/website/src/components/events/virtual-ticket/StickerSelector.tsx [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+                    image: mapStickerIdToPath(stickerPack.stickers, imageId),
+                    imageId,
+                  }))
+                  .sort((a, b) => a.name.localeCompare(b.name))
+                  .map(({ name, disabled, imageId, image }) => (
+                    <li key={imageId} className="flex-shrink-0">
+                      <button
+                        className={`select-none rounded-lg p-0.5 shadow-lg transition-transform ${
+                          disabled ? "bg-gray-300" : "bg-white hover:scale-110"
+                        }`}
```
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/components/events/virtual-ticket/StickerSelector.tsx]: good point

File: apps/website/src/components/events/virtual-ticket/elements/Movable.tsx [hunk 26]
```diff
@@ -0,0 +1,101 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
[the remark this answers is not in the document]
Comment (pjeweb) [reply apps/website/src/components/events/virtual-ticket/elements/Movable.tsx]: no we did not need it at all. good catch

**Review by MattIPv4 — APPROVED** [review]
